### PR TITLE
second attempt at implementing #2404

### PR DESCRIPTION
**Pull request by pzinn**
compared to my first attempt, following @mahrud suggestion:
- no primed functions, fewer changes overall
- no strong dependence on tvalues.m2
cons:
- list of redefined functions is hard-coded (currently in typicalvalues.m2)
- this redefinition needs to be done at the very end (had to move it out of reals.m2, though might eventually want to reorganize a bit)
~Before I continue with this I'd like some feedback on which option is better.~

### CODE REVIEW

File: M2/Macaulay2/m2/typicalvalues.m2 [hunk 1]
```diff
@@ -186,9 +193,150 @@ typval = x -> (
...
+	     if args#0 === RR then f (Constant, args#1) := (x,y) -> f'(numeric x,y);
+	     if args#1 === RR then f (args#0, Constant) := (x,y) -> f'(x,numeric y);
+	     if args === (RR,RR) then f(Constant, Constant) := (x,y) -> f'(numeric x,numeric y); -- phew
+	     ))
      )
-if not member("--no-tvalues", commandLine) then load "tvalues.m2"
+
+load "tvalues.m2"
+
+scan(keys redefs, f -> globalAssign(baseName f,redefs#f));
```
Comment (mahrud) [inline M2/Macaulay2/m2/typicalvalues.m2:212]: You can also do `scanKeys(redefs, (f, m) -> ... )`.

File: M2/Macaulay2/m2/reals.m2 [hunk 2]
```diff
@@ -427,14 +428,19 @@ isReal CC := z -> imaginaryPart z == 0
...
+acosh Constant := acosh @@ numeric
 asinh = method()
 asinh Number := z -> log(z+sqrt(z^2+1))
+asinh Constant := asinh @@ numeric
 atanh = method()
 atanh Number := z -> log((1+z)/(1-z))/2
+atanh Constant := atanh @@ numeric
 acoth = method()
 acoth Number := z -> atanh(1/z)
+acoth Constant := acoth @@ numeric
```
Comment (mahrud) [inline M2/Macaulay2/m2/reals.m2:440]: Are these two not exported?

Comment (pzinn) [reply M2/Macaulay2/m2/reals.m2:440]: looks like it. by two you mean acoth and?

File: M2/Macaulay2/m2/typicalvalues.m2 (hunk 1 above)
Comment (pzinn) [reply M2/Macaulay2/m2/typicalvalues.m2:212]: or even better, `scanPairs`

File: M2/Macaulay2/m2/reals.m2 (hunk 2 above)
Comment (mahrud) [reply M2/Macaulay2/m2/reals.m2:440]: atanh, right above it.

**Comment by pzinn** [comment]
concerning file reorganization, I'm not sure I have the time / energy to do this, but I'm happy to give access to my [pzinn/numeric](https://github.com/pzinn/M2/tree/numeric) branch to anyone that wants to do so.

**Comment by pzinn** [comment]
the test D01.m2 (cf `M2 --check 1`) doesn't seem to like all these statements of the form
```
setupfun("log1p",log1p).Protected=false;
```
not sure what to do about it.

**Comment by DanGrayson** [comment]
> the test D01.m2 (cf `M2 --check 1`) doesn't seem to like all these statements of the form
> 
> ```
> setupfun("log1p",log1p).Protected=false;
> ```
> 
> not sure what to do about it.

I tried it here and it seems okay.

Can you get more information by entering the debugger?

**Comment by pzinn** [comment]
> > the test D01.m2 (cf `M2 --check 1`) doesn't seem to like all these statements of the form
> > ```
> > setupfun("log1p",log1p).Protected=false;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > not sure what to do about it.
> 
> I tried it here and it seems okay.
> 
> Can you get more information by entering the debugger?

It's not easy to debug because `M2 --check 1` does not load `Core`, AFAICT. but I'll try.

**Comment by pzinn** [comment]
I'm afraid I'm stuck on this one. `M2 --check 1` fails for me just like in the unsuccessful builds, and it's hard to diagnose w/o Core being loaded...

**Comment by DanGrayson** [comment]
I'll take a look ...

**Comment by pzinn** [comment]
related to a comment @mahrud made in another thread: is there any reason why actors3.d has so much stuff like
```
atan2(yy:Expr,xx:Expr):Expr := (
     when yy
     is y:RRcell do (
	  when xx
	  is x:RRcell do toExpr(atan2(y.v,x.v))			            -- # typical value: atan2, RR, RR, RR
	  is x:RRicell do toExpr(atan2(toRRi(y.v),x.v))			            -- # typical value: atan2, RR, RRi, RRi
	  is x:ZZcell do toExpr(atan2(y.v,toRR(x.v,precision(y.v))))	    -- # typical value: atan2, RR, ZZ, RR
	  is x:QQcell do toExpr(atan2(y.v,toRR(x.v,precision(y.v))))	    -- # typical value: atan2, RR, QQ, RR
	  else WrongArg(1,"a number"))
     is y:ZZcell do (
	  when xx
	  is x:RRcell do toExpr(atan2(toRR(y.v,precision(x.v)),x.v))    -- -- # typical value: atan2, ZZ, RR, RR
	  is x:RRicell do toExpr(atan2(toRRi(y.v,precision(x.v)),x.v))			            -- # typical value: atan2, ZZ, RRi, RRi
	  is x:ZZcell do toExpr(atan2(toRR(y.v),toRR(x.v)))	       -- # typical value: atan2, ZZ, ZZ, RR
	  is x:QQcell do toExpr(atan2(toRR(y.v),toRR(x.v)))	       -- # typical value: atan2, ZZ, QQ, RR
	  else WrongArg(1,"a number"))
     is y:QQcell do (
	  when xx
	  is x:RRcell do toExpr(atan2(toRR(y.v,precision(x.v)),x.v))    -- # typical value: atan2, QQ, RR, RR
	  is x:RRicell do toExpr(atan2(toRRi(y.v,precision(x.v)),x.v))			            -- # typical value: atan2, QQ, RRi, RRi
	  is x:ZZcell do toExpr(atan2(toRR(y.v),toRR(x.v)))	    -- # typical value: atan2, QQ, ZZ, RR
	  is x:QQcell do toExpr(atan2(toRR(y.v),toRR(x.v)))	    -- # typical value: atan2, QQ, QQ, RR
     	  else WrongArg(1,"a number"))
     is y:RRicell do (
	  when xx
	  is x:RRcell do toExpr(atan2(y.v,toRRi(x.v)))			            -- # typical value: atan2, RRi, RR, RRi
	  is x:RRicell do toExpr(atan2(y.v,x.v))			            -- # typical value: atan2, RRi, RRi, RRi
	  is x:ZZcell do toExpr(atan2(y.v,toRRi(x.v,precision(y.v))))	    -- # typical value: atan2, RRi, ZZ, RRi
	  is x:QQcell do toExpr(atan2(y.v,toRRi(x.v,precision(y.v))))	    -- # typical value: atan2, RRi, QQ, RRi
	  else WrongArg(1,"a number"))
     else WrongArg(2,"a number")
     );
```
i.e., extremely long lists of cases. why don't we do this at the level of .m2 by using M2's inheritance? in exactly the way I handled Constants, one could have a `atan2 (Number,Number) := (x,y) -> atan2(numeric x,numeric y)` and problem solved...

**Comment by DanGrayson** [comment]
I'll try to figure out the `--check 1` problem today, by building on an ubuntu virtual machine here.

**Comment by DanGrayson** [comment]
It looks good to go!

**Comment by pzinn** [comment]
ah! got it! awesome!

**Comment by pzinn** [comment]
I took this opportunity to add two more functions that were implemented by mpfr, namely `Beta` and `Digamma`, and one constant, `CatalanConstant`. added doc as well.

**Comment by pzinn** [comment]
oh, and finally exported `acosh` and `atanh`, as noticed by @mahrud

**Pull request merged**